### PR TITLE
Cache layer extents

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR supersedes #1950 by caching layer extents and then clearing cache only when the dims get updated. The `_update_dims` method needs to get called whenever the data or transforms change, which is precisely when the layer extents need to get recalculated. Putting the clearing of the cache here ensures nothing will get out of sync.

This PR still needs tests.

Thoughts @Czaki @andy-sweet @tlambert03 on this approach?

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by Czaki** [comment]
If you are sure that _update_disms is called on every such case I will be happy with this. But this PR should contain also review and refactor codebase for places where `_update_dims` is called after event execution like here (the only one which I found):

https://github.com/napari/napari/blob/d9da2a9f827632c72eca92f444578a6b9a15acc0/napari/layers/tracks/tracks.py#L381-L383 

to protect taha event executed function will use old extent. 

Is there any possibility to protect from the wrong order of calls in the future?

**Comment by sofroniewn** [comment]
> (the only one which I found):

Fixed

> Is there any possibility to protect from the wrong order of calls in the future?

We could put the call inside a data setter on the base layer and then have each layer provide an `_set_data` method themselves. That would be the safest, but might lead to some breaking changes for people creating their own layers. I think we can do in a followup PR

**Comment by andy-sweet** [comment]
> The `_update_dims` method needs to get called whenever the data or transforms change, which is precisely when the layer extents need to get recalculated.

Yep, like @Czaki, I prefer this PR to #1950, given this is true. I'd be OK with this PR after checking for any cases similar to those @Czaki found and adding a least a little test coverage.

I guess converting the `Layers` to `EventedModels` (or similar) could help protect against some ordering problems in the sense that the event associated with a setter is only emitted after the setter finishes.

**Comment by sofroniewn** [comment]
> I guess converting the Layers to EventedModels (or similar) could help protect against some ordering problems in the sense that the event associated with a setter is only emitted after the setter finishes.

Yes, that would be the hope too

>  I'd be OK with this PR after checking for any cases similar to those @Czaki found and adding a least a little test coverage.

Ok I can add some tests now

**Comment by sofroniewn** [comment]
I have incorporated the suggestion from @andy-sweet on reducing calls to `_extent_data` https://github.com/napari/napari/pull/1950#discussion_r747698274

I have added a minimal test from https://github.com/napari/napari/pull/1950#discussion_r747705728 at the suggestion of @Czaki 

@andy-sweet please let me know if you'd like to see more tests here. What i've done barely scratches the surface I feel. If you'd like to submit those tests as a PR to this PR feel free. Or if it looks good to you as it is now then that's fine too. Thanks!!

File: napari/layers/base/base.py [hunk 1]
```diff
@@ -698,12 +700,14 @@ def _get_extent_world(self, data_extent):
     @property
     def extent(self) -> Extent:
         """Extent of layer in data and world coordinates."""
-        data = self._extent_data
-        return Extent(
-            data=data,
-            world=self._get_extent_world(data),
-            step=abs(self._data_to_world.scale),
-        )
+        extent_data = self._extent_data
```
Comment (tlambert03) [inline napari/layers/base/base.py:692]: doesn't have to be done here, but we really ought to use `functools.cached_property`, (and vendor it from functools for py<3.8).  This is exactly what it does, with a slightly cleaner API for invalidating the cache

Comment (tlambert03) [reply napari/layers/base/base.py:692]: actually, it also would mean we don't need `self._extent_cache = None` above.  which is a bit nicer.  ~@Czaki would you want to do that here?  Or prefer to leave it for elsewhere~

sorry... thought this was @Czaki's PR... @sofroniewn ... what do you think?

File: napari/layers/_tests/test_layer_attributes.py [hunk 2]
```diff
@@ -18,3 +19,24 @@ def test_layer_bounding_box_order(image_shape, dims_displayed, expected):
...
+    layer.opacity = 1
+    old_extent = layer.extent
+    assert old_extent is layer.extent
+    layer.opacity = 0.5
+    assert old_extent is layer.extent
+
+    # Check layer extent change when scale changes
+    layer.scale = (2,) + layer.scale[1:]
+    old_extent2 = layer.extent
+    assert old_extent is not layer.extent
```
Comment (andy-sweet) [inline napari/layers/_tests/test_layer_attributes.py:55]: I still don't love these assertions because they're testing an implementation detail that could change. This PR is an optimization that is intended not to change the functional behavior of napari. The tests act as a defense to make sure that functional behavior is maintained regardless of how we choose to implement how `layer.extent` gives us the right value.

Ideally, I think we'd assert that the extent has the expected value based on the attribute we update (e.g. scale), but that's not easy

The test name at least does reflect that this is testing the cache as an implementation detail, so I don't think it's a blocker.

File: napari/layers/base/base.py (hunk 1 above)
Comment (Czaki) [reply napari/layers/base/base.py:692]: But how did you would like to invalidate `functools.cached_property` on the update of property like `scale`, `translate` or updating data? 

`self._extent_cache` is settled to None on `_update_dims` call.

Comment (tlambert03) [reply napari/layers/base/base.py:692]: is the question how does one invalidate a `cached_property`?  If so, you do it by deleting the attribute:
 https://docs.python.org/3/library/functools.html#functools.cached_property

so `self._extent_cache = None` would be the same as `del self.extent`

File: napari/layers/_tests/test_layer_attributes.py [hunk 3]
```diff
@@ -18,3 +19,24 @@ def test_layer_bounding_box_order(image_shape, dims_displayed, expected):
     assert np.allclose(
         layer._display_bounding_box(dims_displayed=dims_displayed), expected
     )
+
+
+@pytest.mark.parametrize('Layer, data, ndim', layer_test_data)
```
Comment (andy-sweet) [inline napari/layers/_tests/test_layer_attributes.py:24]: I guess we could also parameterize by the attribute we update (e.g. `scale`, `data`), but maybe that's overkill and could be hard to follow.

File: napari/layers/base/base.py (hunk 1 above)
Comment (Czaki) [reply napari/layers/base/base.py:692]: I miss this sentence when read bout` cache_property`. 

When is a plan to drop python 3.7?

Comment (tlambert03) [reply napari/layers/base/base.py:692]: end of the year... but that really doesn't need to come into play in our decision here. It's a tiny little bit of code to vendor

File: napari/layers/base/base.py [hunk 4]
```diff
@@ -698,12 +700,14 @@ def _get_extent_world(self, data_extent):
...
+        if self._extent_cache is None:
+            self._extent_cache = Extent(
+                data=extent_data,
+                world=self._get_extent_world(extent_data),
+                step=abs(self.scale),
```
Comment (andy-sweet) [inline napari/layers/base/base.py]: This represents a change: `self.scale` (which you could maybe think of as pixel spacing) is not always equal to `self._data_to_world.scale` (which represents the scale of two composed transforms).

If I understand `extent` correctly (that's not a given yet), then I think you want `self._data_to_world.scale`.

File: napari/layers/base/base.py (hunk 1 above)
Comment (sofroniewn) [reply napari/layers/base/base.py:692]: I support vendor - I am not at my laptop for a while, so if someone wants to do this, that’s fine by me, otherwise I can do it tonight

File: napari/layers/base/base.py [hunk 5]
```diff
@@ -698,12 +700,14 @@ def _get_extent_world(self, data_extent):
...
-        data = self._extent_data
-        return Extent(
-            data=data,
-            world=self._get_extent_world(data),
-            step=abs(self._data_to_world.scale),
-        )
+        extent_data = self._extent_data
+        if self._extent_cache is None:
+            self._extent_cache = Extent(
+                data=extent_data,
```
Comment (andy-sweet) [inline napari/layers/base/base.py]: [My comment in the previous PR](https://github.com/napari/napari/pull/1950#discussion_r747698274) was actually related to storing the result of `self._data_to_world`, not `self._extent_data`, but I guess we could store both?

**Review by andy-sweet — CHANGES_REQUESTED** [review]
The only thing that needs to be changed or clarified is the use of `self.scale` in `Layer.extent`. The other comments are optional.

Regarding tests, I'm happy to try to add some more in a follow-up PR - by my comments, I don't think it's straightforward to improve them.

File: napari/layers/base/base.py (hunk 1 above)
Comment (grlee77) [reply napari/layers/base/base.py:692]: nice, I was not aware of `functools.cached_property`. @jni: we should keep this in mind as a future update to regionprops caching for scikit-image

**Comment by grlee77** [comment]
> @andy-sweet please let me know if you'd like to see more tests here. What i've done barely scratches the surface I feel. If you'd like to submit those tests as a PR to this PR feel free. Or if it looks good to you as it is now then that's fine too. Thanks!!

I see you have tested a change to the layer's transform. To cover a case modifying the data, perhaps adding new points to a points layer would be a good one to have

File: napari/_vendor/cpython/functools.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
+from _thread import RLock
+
+################################################################################
+### cached_property() - computed once per instance, cached as attribute
+################################################################################
+
```
Comment (tlambert03) [inline napari/_vendor/cpython/functools.py:6]: maybe add:

```py
try:
    from functools import cached_property
except ImportError:
    ...

**Comment by sofroniewn** [comment]
I still have to improve tests, but @tlambert03 is this what you were expecting to see with the vendored `cached_property`?

If i used `del` i got the following

```python-traceback
  File "/Users/nsofroniew/GitHub/napari/napari/layers/base/base.py", line 643, in _update_dims
    del self.extent
AttributeError: extent
```

so i had to look inside `self.__dict__` (base on this [SO post](https://stackoverflow.com/questions/59899732/python-cached-property-how-to-delete)

File: napari/_vendor/cpython/functools.py (hunk 6 above)
Comment (sofroniewn) [reply napari/_vendor/cpython/functools.py:6]: just adjusted

File: napari/layers/base/base.py [hunk 7]
```diff
@@ -640,7 +640,8 @@ def _update_dims(self, event=None):
             self._position = (0,) * (ndim - old_ndim) + self._position
 
         self._ndim = ndim
-        self._extent_cache = None
+        if 'extent' in self.__dict__:
+            del self.__dict__['extent']
```
Comment (tlambert03) [inline napari/layers/base/base.py]:
```suggestion
            del self.extent
```

(the documented way to invalidate cached_property)

**Review by tlambert03 — APPROVED** [review]

File: napari/_vendor/cpython/functools.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
+from _thread import RLock
```
Comment (tlambert03) [inline napari/_vendor/cpython/functools.py]: since you're not bringing in `_thread`, this probably needs to be `from threading import RLock`

**Comment by tlambert03** [comment]
> I still have to improve tests, but @tlambert03 is this what you were expecting to see with the vendored cached_property?
> If i used del i got the following

yep! looks good.  yeah, looking in `self.__dict__` seems like the right way.  my suggestion above is negligible

File: napari/_vendor/cpython/functools.py (hunk 8 above)
Comment (sofroniewn) [reply napari/_vendor/cpython/functools.py]: changed

File: napari/layers/base/base.py (hunk 7 above)
Comment (sofroniewn) [reply napari/layers/base/base.py]: changed

File: napari/layers/base/base.py (hunk 1 above)
Comment (sofroniewn) [reply napari/layers/base/base.py:692]: now using `cached_property`

File: napari/layers/base/base.py (hunk 5 above)
Comment (sofroniewn) [reply napari/layers/base/base.py]: ohh i got you now - yup read that too quickly, have fixed

**Comment by sofroniewn** [comment]
I've made some updates, but I'm now getting all sorts of test failures like 

```python-traceback
napari/layers/image/_tests/test_image.py ...........................................................F..................................................................                              [100%]

================================================================================================= FAILURES =================================================================================================
__________________________________________________________________________________________ test_world_data_extent __________________________________________________________________________________________

    def test_world_data_extent():
        """Test extent after applying transforms."""
        np.random.seed(0)
        shape = (6, 10, 15)
        data = np.random.random(shape)
        layer = Image(data)
        extent = np.array(((0,) * 3, shape))
>       check_layer_world_data_extent(layer, extent, (3, 1, 1), (10, 20, 5), True)

napari/layers/image/_tests/test_image.py:669:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

layer = <Image layer 'Image' at 0x7ff2cdb5e7d0>, extent = array([[ 0,  0,  0],
       [ 6, 10, 15]]), scale = (3, 1, 1), translate = (10, 20, 5), pixels = True

    def check_layer_world_data_extent(
        layer, extent, scale, translate, pixels=False
    ):
        """Test extents after applying transforms.

        Parameters
        ----------
        layer : napari.layers.Layer
            Layer to be tested.
        extent : array, shape (2, D)
            Extent of data in layer.
        scale : array, shape (D,)
            Scale to be applied to layer.
        translate : array, shape (D,)
            Translation to be applied to layer.
        """
        np.testing.assert_almost_equal(layer.extent.data, extent)
        world_extent = extent - 0.5 if pixels else extent
>       np.testing.assert_almost_equal(layer.extent.world, world_extent)
E       AssertionError:
E       Arrays are not almost equal to 7 decimals
E
E       Mismatched elements: 6 / 6 (100%)
E       Max absolute difference: 0.5
E       Max relative difference: 1.
E        x: array([[ 0.,  0.,  0.],
E              [ 6., 10., 15.]])
E        y: array([[-0.5, -0.5, -0.5],
E              [ 5.5,  9.5, 14.5]])

napari/_tests/utils.py:198: AssertionError
```

Does @grlee77 or @andy-sweet understand why? It's not clear to me looking at the diff what I changed that would cause this.

I've tried to improve the tests a little, but i totally agree with @andy-sweet they aren't perfect

File: napari/layers/_tests/test_layer_attributes.py (hunk 2 above)
Comment (sofroniewn) [reply napari/layers/_tests/test_layer_attributes.py:55]: 100% agreed, i've tried to improve, and we can merge as is now and then update later or we can change now. i really don't mind

Comment (andy-sweet) [reply napari/layers/_tests/test_layer_attributes.py:55]: I like the changes and I think these are good enough to merge now. We can think about better coverage through parameterization later.

**Review by andy-sweet — APPROVED** [review]
LGTM!

**Comment by andy-sweet** [comment]
> Does @grlee77 or @andy-sweet understand why?

Looks like `ImageBase` overrode `_get_extent_world`, but that's no longer defined on the base `Layer` and therefore isn't used any more. But it handles that 0.5 offset that the test is looking for. So we'll either have bring it back into the base layer or do some other refactoring.

It would kind of nice to have something [like Java's @Override decorator](https://docs.oracle.com/javase/8/docs/api/java/lang/Override.html) to flag these kinds of errors. Fortunately, we had tests to catch this one, but we might not always be so lucky. Not sure if something standard like that exists for Python or if we could easily roll our own.

**Comment by sofroniewn** [comment]
> Looks like ImageBase overrode _get_extent_world, but that's no longer defined on the base Layer and therefore isn't used any more. But it handles that 0.5 offset that the test is looking for. So we'll either have bring it back into the base layer or do some other refactoring.

Ah nice, good catch!! I was really puzzled last night, that's the only thing that makes sense!

I have fixed this now by adding an `_array_like` attribute. I thought we used to have one, but we must have gotten rid of it. This might not be the cleanest solution but it gets the job done.

Maybe @andy-sweet you can take one more look. If you're good with changes and tests are passing I can then merge. Thanks!!

**Review by andy-sweet — APPROVED** [review]
LGTM x2

**Review by Czaki — CHANGES_REQUESTED** [review]
I would like to see the test for all parameters (to be sure that some overwritten does not ommits calling of `_update_dims`. And also check if `_update_dims` is called before the signal call. 

I would like to see "this function should be called before emit signal informing about parameter change` in docstring of `_update_dims`

**Comment by sofroniewn** [comment]
> I would like to see "this function should be called before emit signal informing about parameter changein docstring of_update_dims`

I have updated the doc string

> I would like to see the test for all parameters (to be sure that some overwritten does not ommits calling of _update_dims. And also check if _update_dims is called before the signal call.

That's a good idea, but I think can come in a follow up PR. I think also more important to test the actual behavior is what we want rather than an explicit call to a private method, which is definitely an implementation detail. Ideally we'd move over to using our evented models for layers and we could find a way to trigger the update more automatically.

I will merge this once tests are passing. Thanks!!

**Review by Czaki — APPROVED** [review]
